perceval/simulators/feed_forward_simulator.py

```diff
@@ -0,0 +1,296 @@
+
+            # Use only the first mode as circuits can have different sizes
+            components[i] = (circ_r[0], config.default_circuit)
+
+        # We can't reject any state at this moment since we need all possible measured states
+        sim = self._init_simulator(input_state, components)
+
+        # Estimation of possible measures: n for each measured mode
+        n = input_state.n if isinstance(input_state, BasicState) else input_state.n_max
+        intermediate_progress = 1 / n ** len(measured_modes)
+        prog_cb = partial_progress_callable(progress_callback, max_val=intermediate_progress)
+
+        if is_svd:
+            default_res = sim.probs_svd(input_state, detectors, progress_callback=prog_cb, normalize=False)["results"]
+        else:
+            default_res = sim.probs(input_state, normalize=False)
+
+        prog_cb = partial_progress_callable(progress_callback, min_val=intermediate_progress)
+
+        # 3: deduce all measurable states and launch one simulation for each of them
+        res = BSDistribution()
+        simulated_measures = set()
+
+        global_perf = 0  # = P(logic_filter and n >= photon_filter)
+
+        for j, (state, prob) in enumerate(default_res.items()):
+            sub_circuits = [config.configure(state[slice(self._components[i][0][0], self._components[i][0][-1] + 1)])
+                            for i, config in considered_config]
+
+            if all(c1 == default_circuit for c1, default_circuit in zip(sub_circuits, default_circuits)):
+                # Default case, no need for further computation
+
+                # There is a problem here: if the same circuit is instanced twice, the == operator returns False
+                # This can be a problem if we copy the default circuit or instantiate a new one
+
+                if self._post_process_state(state):
+                    res[state] = prob
+                    global_perf += prob
+
+                if prog_cb is not None:
+                    prog_cb((j + 1) / len(default_res), "Computing feed-forwarded circuit")
+                continue
+
+            measured_state = tuple(state[i] for i in measured_modes)
+
+            if measured_state not in simulated_measures:
+
+                for sub_i, (i, config) in enumerate(considered_config):
+                    r = self._components[i][0]
+                    # Does not use the list of modes as circuits can have different sizes
+                    components[i] = (config.config_modes(r)[0], sub_circuits[sub_i])
+
+                new_heralds = {i: state[i] for i in measured_modes}
+                sim = self._init_simulator(input_state, components, filter_states=True, new_heralds=new_heralds)
+
+                if is_svd:
+                    new_prog_cb = partial_progress_callable(prog_cb, j / len(default_res), (j + 1) / len(default_res))
+                    sub_res = sim.probs_svd(input_state, detectors, new_prog_cb, normalize=False)["results"]
```

Comment:
When the FF configurator injects a `Processor` with heralds, the `input_state` here needs to be adjusted to take the modes covered by heralds into account.